File: problems-extend-adapt/1401-1500/1474_prune-a-linked-list-in-keep-drop-cycles/solution_values.ts
function pruneNodes(head: ListNode | null, m: number, n: number): ListNode | null {
    // Sweep 1: record the values that survive each keep-m, drop-n cycle.
    // A keep run cut short by the tail simply ends the walk; a drop run
    // steps past the nodes it loses.
    const kept: number[] = [];
    let node = head;
    while (node !== null) {
        for (let i = 0; i < m && node !== null; ++i) {
            kept.push(node.val);
            node = node.next;
        }
        for (let i = 0; i < n && node !== null; ++i) {
            node = node.next;
        }
    }
    // Sweep 2: rebuild a fresh list threaded from the surviving values.
    const dummy = new ListNode(0);
    let tail = dummy;
    for (const value of kept) {
        tail.next = new ListNode(value);
        tail = tail.next;
    }
    return dummy.next;
}

File: problems-originals/1401-1500/1474_delete-n-nodes-after-m-nodes-of-a-linked-list/solution_values.py
class Solution:
    def deleteNodes(self, head: Optional[ListNode], m: int, n: int) -> Optional[ListNode]:
        # Sweep 1: record the values that survive each keep-m, drop-n cycle.
        # A keep run cut short by the tail simply ends the walk; a drop run
        # steps past the nodes it loses.
        kept = []
        node = head
        while node is not None:
            for _ in range(m):
                kept.append(node.val)
                node = node.next
                if node is None:
                    break
            for _ in range(n):
                if node is None:
                    break
                node = node.next
        # Sweep 2: rebuild a fresh list threaded from the surviving values.
        dummy = ListNode(0)
        tail = dummy
        for value in kept:
            tail.next = ListNode(value)
            tail = tail.next
        return dummy.next
